### PR TITLE
TIP-1285: product model remove subscriber

### PR DESCRIPTION
**Pull request by nmarniesse**
**Description (for Contributor and Core Developer)**

Part of TIP-1285  

**Definition Of Done (for Core Developer only)**

| Q                                 | A
| --------------------------------- | ---
| Added Specs                       | OK
| Added legacy Behats               | -
| Added acceptance tests            | -
| Added integration tests           | -
| Changelog updated                 | OK
| Review and 2 GTM                  | Todo
| Micro Demo to the PO (Story only) | -
| Migration script                  | -
| Tech Doc                          | -

`Todo`: Pending / Work in progress
`OK`: Done / Validated
`-`: Not needed

### CODE REVIEW

File: src/Akeneo/Pim/Enrichment/Bundle/EventSubscriber/ComputeDescendantCompletenessAndIndexProductAndAncestorsSubscriber.php [hunk 1]
```diff
@@ -8,16 +8,21 @@
...
- * Orchestrator for below jobs:
+ * Orchestrator for below jobs on update:
  *  - Computes and saves the completenesses for the variant products of the given product models.
  *  - Indexes these variant products
  *  - Indexes the product models impacted by the new completeness (= ancestors and descendants of the given product models)
  *
+ * On delete:
+ *   - Remove given product models and its descendants (product models and variant products)
+ *   - Re-index ancestor if any
+ *
```
Comment (ahocquard) [inline src/Akeneo/Pim/Enrichment/Bundle/EventSubscriber/ComputeDescendantCompletenessAndIndexProductAndAncestorsSubscriber.php]: :+1:

File: src/Akeneo/Pim/Enrichment/Bundle/Product/Query/Sql/GetDescendantProductModelIds.php [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+namespace Akeneo\Pim\Enrichment\Bundle\Product\Query\Sql;
+
+use Doctrine\DBAL\Connection;
+
+/**
+ * @author    Nicolas Marniesse <nicolas.marniesse@akeneo.com>
+ * @copyright 2019 Akeneo SAS (http://www.akeneo.com)
+ * @license   http://opensource.org/licenses/osl-3.0.php  Open Software License (OSL 3.0)
+ */
+class GetDescendantProductModelIds
```
Comment (ahocquard) [inline src/Akeneo/Pim/Enrichment/Bundle/Product/Query/Sql/GetDescendantProductModelIds.php]: final, and interface if you can't mock it in my opinion. You can put the interface in the component/
product/query
https://matthiasnoback.nl/2018/09/final-classes-by-default-why/

File: src/Akeneo/Pim/Enrichment/Bundle/Elasticsearch/Indexer/ProductModelDescendantsAndAncestorsIndexer.php [hunk 3]
```diff
@@ -70,4 +82,33 @@ public function indexFromProductModelCodes(array $productModelCodes): void
...
+     *
+     * @param array $productModelIds
+     */
+    public function removeFromProductModelIds(array $productModelIds): void
+    {
+        if (empty($productModelIds)) {
+            return;
+        }
+
+        $productIds = $this->getDescendantVariantProductIds->fromProductModelIds($productModelIds);
```
Comment (mmetayer) [inline src/Akeneo/Pim/Enrichment/Bundle/Elasticsearch/Indexer/ProductModelDescendantsAndAncestorsIndexer.php]: won't work on POST_SAVE :/

Comment (nmarniesse) [reply src/Akeneo/Pim/Enrichment/Bundle/Elasticsearch/Indexer/ProductModelDescendantsAndAncestorsIndexer.php]: Thanks! The queries as useless as the cascade removing is already done by ProductModelIndexer.  
So I remove this code

File: src/Akeneo/Pim/Enrichment/Bundle/Product/Query/Sql/GetDescendantProductModelIds.php (hunk 2 above)
Comment (nmarniesse) [reply src/Akeneo/Pim/Enrichment/Bundle/Product/Query/Sql/GetDescendantProductModelIds.php]: The class is now deleted

File: src/Akeneo/Pim/Enrichment/Bundle/EventSubscriber/ComputeDescendantCompletenessAndIndexProductAndAncestorsSubscriber.php [hunk 4]
```diff
@@ -76,15 +83,39 @@ public function fromProductModelsEvent(Event $event): void
...
+    }
+
+    public function fromProductModelsRemoveEvent(RemoveEvent $event): void
+    {
+        $productModels = $event->getSubject();
+        if (!is_array($productModels) || !current($productModels) instanceof ProductModelInterface) {
+            return;
+        }
+
+        $this->productModelDescendantsAndAncestorsIndexer->removeFromProductModelIds($event->getSubjectId());
```
Comment (ahocquard) [inline src/Akeneo/Pim/Enrichment/Bundle/EventSubscriber/ComputeDescendantCompletenessAndIndexProductAndAncestorsSubscriber.php]: don't hesitate to put `$event->getSubjectId()` in a variable to have a name such as `$productModelIds = $event->getSubjectId()`. It's a little bit more explicit in my opinion.

File: src/Akeneo/Pim/Enrichment/Bundle/Product/Query/Sql/GetAncestorAndDescendantProductModelCodes.php [hunk 5]
```diff
@@ -51,4 +51,22 @@ public function fromProductModelCodes(array $productModelCodes): array
             ['codes' => Connection::PARAM_STR_ARRAY]
         )->fetchAll(\PDO::FETCH_COLUMN, 0);
     }
+
+    public function getOnlyAncestorsFromProductModelIds(array $productModelIds): array
```
Comment (ahocquard) [inline src/Akeneo/Pim/Enrichment/Bundle/Product/Query/Sql/GetAncestorAndDescendantProductModelCodes.php]: Maybe part of another class, we will see if we refactor it

**Pull request merged**